In [2]:
import sys
import os

parent_dir = os.path.abspath(os.path.join(os.getcwd(), '..'))
sys.path.append(parent_dir)

from python_utils import *
from tqdm.notebook import tqdm, trange

### Part 1: compute log-volumes

In [ ]:
e0 = 12
def get_ramification_dataset1(l):
    ep_range = dict()
    ep_range['general'] = {'good':[1], 'multi':divisors(12*l)}
    ep_range[2] = {'good':[e for e in divisors(pow(2,8) * pow(3,2)) if e%2 == 0], 'multi':[e for e in divisors(24*l) if e%2 == 0]}
    ep_range[3] = {'good':[e for e in divisors(pow(2,6) * pow(3,2)) if e%2 == 0], 'multi':[e for e in divisors(48*l) if e%2 == 0]}
    ep_range[l] = {'good':[l-1,l*(l-1),l*l-1], 'multi':[(l-1)*e for e in divisors(12*l)]}
    return ep_range

# compute log-volume
my_log_volumes1 = dict()
for p in tqdm(primes(150)):
    if p >= 5 and p <= 30:
        my_log_volumes1[p] = compute_log_volume(get_ramification_dataset1, e0, p)
        print(p, my_log_volumes1[p])
    elif p > 30:
        my_log_volumes1[p] = compute_log_volume_fast(get_ramification_dataset1, e0, p)

NameError: name 'tqdm' is not defined

In [9]:
e0 = 4
def get_ramification_dataset2(l):
    ep_range = dict()
    ep_range['general'] = {'good':[1], 'multi':divisors(4*l)}
    ep_range[2] = {'good':[e for e in divisors(96) if e%2 == 0], 'multi':[e for e in divisors(8*l) if e%2 == 0]}
    ep_range[3] = {'good':[e for e in divisors(12) if e%2 == 0], 'multi':[2*e for e in divisors(8*l)]}
    ep_range[l] = {'good':[l-1,l*(l-1),l*l-1], 'multi':[(l-1)*e for e in divisors(4*l)]}

    return ep_range

# compute log-volume
my_log_volumes2 = dict()
for p in tqdm(primes(150)):
    if p >= 5 and p <= 30:
        my_log_volumes2[p] = compute_log_volume(get_ramification_dataset2, e0, p)
        print(p, my_log_volumes2[p])
    elif p > 30:
        my_log_volumes2[p] = compute_log_volume_fast(get_ramification_dataset2, e0, p)

  0%|          | 0/35 [00:00<?, ?it/s]

5 59.17843935222741
7 50.356443373450205
11 51.87927802801745
13 54.9948057061493
17 61.945242964890056
19 65.7568310535161
23 73.48046840034378
29 85.47170755570102


In [9]:
e0 = 4
def get_ramification_dataset3(l):
    ep_range = dict()
    ep_range['general'] = {'good':[1], 'multi':divisors(4*l)}
    ep_range[2] = {'good':[e for e in divisors(96) if e%2 == 0], 'multi':[e for e in divisors(8*l) if e%2 == 0]}
    ep_range[3] = {'good':[e for e in divisors(12) if e%2 == 0], 'multi':[2*e for e in divisors(8*l)]}
    ep_range[l] = {'good':[], 'multi':[(l-1)*e for e in divisors(4*l)]}
    return ep_range

# compute log-volume
my_log_volumes3 = dict()
for p in tqdm(primes(150)):
    if p >= 5 and p <= 30:
        my_log_volumes3[p] = compute_log_volume(get_ramification_dataset3, e0, p)
        print(p, my_log_volumes3[p])
    elif p > 30:
        my_log_volumes3[p] = compute_log_volume_fast(get_ramification_dataset3, e0, p)

  0%|          | 0/35 [00:00<?, ?it/s]

5 59.17843935222741
7 50.356443373450205
11 51.519593737097686
13 54.466099233201966
17 61.266151118499025
19 65.0109420729845
23 72.71375032818264
29 84.62504794610089


### Part 2: upper bounds for $r_3$, $s_3$, $r_l$, $s_l$ 

In [169]:
my_primes = [11,17,19,23,29,31,37,41]
a1_dict = dict()
a2_dict = dict()
for p in my_primes:
    a1_dict[p] = 6*(p*p+5*p)/(p*p+p-12)*(1-1/4/p)
    a2_dict[p] = max(6*my_log_volumes2[p]+ a1_dict[p]*log(2*3) , 6*my_log_volumes3[p] + a1_dict[p]*log(2*3*p))
    print(p, a1_dict[p], a2_dict[p])

11 8.600000000000001 345.14859320401337
17 7.520408163265307 402.3785900108973
19 7.33695652173913 424.8149345321058
23 7.0777777777777775 471.15650860737514
29 6.835664335664336 543.0158579911662
31 6.777551020408164 577.3752837869084
37 6.643472022955523 648.8286584098873
41 6.577192982456141 697.3895048776211


A computation in the proof of upper bounds for $r_l$, $s_l$.


In [170]:
# upper bounds for r_l, s_l
for p in my_primes:
    print(p, (a2_dict[p] / log(p) + 7.6) / 3, (a2_dict[p] / log(p) + 7.6) / (20 / 7)) 

11 50.512714406861356 53.03835012720442
17 49.87399581880474 52.36769560974498
19 50.625679338566705 53.15696330549504
23 52.62182296853172 55.2529141169583
29 56.287234129683156 59.10159583616731
31 58.57851244104773 61.50743806310011
37 62.42841394777644 65.54983464516526
41 65.13159832374322 68.38817823993038


About the proof of upper bounds for $r_3$, $s_3$.

Suppose that $3 \mid x$. 

Let $q\ge 17$ be the smallest prime number, such that $q\nmid rs$, $q\nmid (3r_3r-3)$.

Then we have $ (3r_3r-3-a_1(l))\cdot \log(3) \le a_2(q)$.

Hence $r_3r \le \frac{a_2(q)/\log(3)+3+a_1(l)}{3}$.

As a vary rough upper bound, we have $r_3 r < 1000$.

Then we can search for all possible $(r_3, r)$.

Note that by replacing $s$ by its prime divisors $\ge 11$, we can assume that $s$ excludes at most one possibel $p\ge 17$.

In [171]:
max_r = 1300 # 7 <= r,s <= 1226

r3_dict = dict()
for r in trange(7,max_r+1):
    r3_dict[r] = 0
    for r3 in range(1, 1000 // r + 1):
        possible_q = []
        for q in [17,19,23,29,31,37,39,41]:
            if r%q != 0 and (r*r3-1) % q != 0:
                possible_q.append(q) # one possible q for the case s%q == 0
            if len(possible_q) == 2:
                break
        if (len(possible_q) < 2):
            print(f'Error for r={r}, r3={r3}!')
        if r3*r > int((a2_dict[q] / log(3) + 3 + a1_dict[q])/3):
            # beyond the bound
            continue
        else:
            # maximal possible r * r3
            r3_dict[r] = r * r3 
    # since 3^r = (3^{r/r1})^r1
    r3_dict[r] = min([r3_dict[r1] for r1 in divisors(r) if r1 >= 7])
    if r3_dict[r] > 133:
        print(r, r3_dict[r])



  0%|          | 0/1294 [00:00<?, ?it/s]

17 153
34 136
51 153
67 134
68 136
134 134
137 137


Hence we have $r \cdot r_3 \le 153$; since $r > r \cdot r_3$, we have $r_3 = 0$ when $r > 137$. 

If $r\ge 8$, then we have $s\cdot s_3 \le 153$; $s_3 = 0$ when $s > 137$ by symmetry.

If $r = 7$, then $l\nmid r$, we can compute similarly.

In [172]:
max_r = 1300 # r = 7,  11 <= s <= 1226

s3_dict = dict()
for s in trange(11,max_r+1):
    s3_dict[s] = 0
    for s3 in range(1, 1000 // s + 1):
        for q in [17,19,23,29,31,37,39,41]:
            if s%q != 0 and (s*s3-1) % q != 0 and (s*s3-3) % q != 0:
                # the conditions based on assersion (iii) are different!
                break
                
        if not(s%q != 0 and (s*s3-1) % q != 0 and (s*s3-3) % q != 0):
            print(f'Error for s={s}, s3={s3}!')
        if s3*s > int((a2_dict[q] / log(3) + 3 + a1_dict[q]) / (20 / 7)):
            # beyond the bound
            continue
        else:
            # maximal possible s * s3
            s3_dict[s] = s * s3 
    # since 3^s = (3^{s/s1})^s1
    if s%2 == 0 or s%3 == 0 or s%7 == 0:
        # proved signatures; (2,3,7), (3,3,7), (3,7,7)
        s3_dict[s] = 0
    s3_dict[s] = min([s3_dict[s1] for s1 in divisors(s) if s1 >= 11])
    if s3_dict[s] > 130:
        print(s, s3_dict[s])



  0%|          | 0/1290 [00:00<?, ?it/s]

17 153
131 131
137 137


Hence we have $s \cdot s_3 \le 153$; $s_3 = 0$ when $s > 137$. 

### Part 3: upper bounds for $r_2$, $s_2$

In [181]:
my_primes = [11,17,19,23,29,31,37,41]
a1_dict = dict()
a2_dict = dict()
for p in my_primes:
    a1_dict[p] = 6*(p*p+5*p)/(p*p+p-12)*(1-1/12/p)
    a2_dict[p] = 6*my_log_volumes1[p] + a1_dict[p]*log(2*3*p)
    print(p, a1_dict[p], a2_dict[p])

11 8.733333333333334 796.5615288095332
17 7.595238095238096 977.1608010046082
19 7.4021739130434785 1040.7370146010778
23 7.12962962962963 1177.0895285582615
29 6.875291375291375 1388.1992527100615
31 6.814285714285715 1474.5661407635166
37 6.673601147776184 1687.100875332116
41 6.604093567251462 1832.0968204893907


In [199]:
f"{exp(472 / (3 * 13 -7)):.2e}", 472/(3*138-7.6),log(4)

('2.55e+06', 1.1614173228346458, 1.3862943611198906)